### PR TITLE
TST: query with timezone aware index & column

### LINKED ISSUE #29463: "Cannot compare tz-naive and tz-aware datetime-like objects" when querying DatetimeIndex
## Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({
    'val': range(10),
    'time': pd.date_range(start='2019-01-01', freq='1d', periods=10, tz='UTC')
})

# This works.
df.query('"2019-01-03 00:00:00+00" < time')

# This raises an exception.
df.set_index('time').query('"2019-01-03 00:00:00+00" < time')

# Perhaps interestingly, this works.
df.set_index('time').reset_index().query('"2019-01-03 00:00:00+00" < time')

```
#### Problem description

query shouldn't behave differently when the thing being queried is an index vs a column. Also, the error just seems objectively incorrect, as both values are tz-aware to start with (perhaps it's getting lost along the way).

May be related to #26236 , but the repro is different.

#### Expected Output
Not an exception.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7.final.0
python-bits: 64
OS: Linux
OS-release: 4.3.5-smp-821.23.0.0
machine: x86_64
processor: 
byteorder: little
LC_ALL: en_US.UTF-8
LANG: None
LOCALE: en_US.UTF-8

pandas: 0.24.1
pytest: None
pip: None
setuptools: unknown
Cython: None
numpy: 1.16.4
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 2.0.0
sphinx: None
patsy: 0.4.1
dateutil: 2.8.0
pytz: 2019.3
blosc: None
bottleneck: None
tables: 3.5.2
numexpr: 2.6.10dev0
feather: None
matplotlib: 3.0.3
openpyxl: None
xlrd: 1.2.0
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10.3
s3fs: None
fastparquet: None
pandas_gbq: 0+unknown
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by vipulr8**
- [X] closes #29463
- [X] tests added / passed
- [X] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/indexing/test_column_vs_index_tz.py
Comment (MarcoGorelli) [inline pandas/tests/frame/indexing/test_column_vs_index_tz.py:53]: Thanks @vipulrai91 

There's a fixture you can use here, `tz_aware_fixture` (feel free to ask for help if it's unclear how to use it)

Comment (vipulrai91) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:53]: @MarcoGorelli thank you for the feedback.

As far as I understand the change has to be.        

```
def test_check_column_vs_index_tz_query(self, tz_aware_fixture):
    tz_list = tz_aware_fixture
```
Got this snippet from test_datetime.py , but what actually is tz_aware_fixture?
Also , How Do I run the test locally, I tried going through the docs , but will need some help here.

Thanks

Comment (MarcoGorelli) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:53]: `tz = tz_aware_fixture` should be enough, then pytest will loop through the different timezones defined in `tz_aware_fixture`

> what actually is tz_aware_fixture?
It's a [fixture](https://docs.pytest.org/en/latest/fixture.html) which you can use to test different timezones.

> How Do I run the test locally

See [using pytest](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#using-pytest). So here,
```bash
$ pytest pandas/tests/frame/indexing/test_column_vs_index_tz.py
```
should be enough

Comment (vipulrai91) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:53]: Thank you for guiding.

tz_aware_fixture is a combination of TimeZones and other object types , also even 
it throws pytz.exceptions.UnknownTimeZone error.



```
test_column_vs_index_tz.py:58: TypeError
============================================================ short test summary info =============================================================
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query['UTC'] - pytz.exceptions.UnknownTimeZoneErr...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query['US/Eastern'] - pytz.exceptions.UnknownTime...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query['Asia/Tokyo'] - pytz.exceptions.UnknownTime...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query['dateutil/US/Pacific'] - pytz.exceptions.Un...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query['dateutil/Asia/Singapore'] - pytz.exception...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[tzutc()] - TypeError: 'tzutc' object is not...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[tzlocal()] - TypeError: 'tzlocal' object is...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[pytz.FixedOffset(300)] - TypeError: '_Fixed...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[<UTC>] - TypeError: 'UTC' object is not ite...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[pytz.FixedOffset(-300)] - TypeError: '_Fixe...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[datetime.timezone.utc] - TypeError: 'dateti...
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[datetime.timezone(datetime.timedelta(seconds=3600))]
FAILED test_column_vs_index_tz.py::TestColumnvsIndexTZEquality::test_check_column_vs_index_tz_query[datetime.timezone(datetime.timedelta(days=-1, seconds=82800), 'foo')]
```

File: pandas/tests/frame/indexing/test_column_vs_index_tz.py
Comment (mroeschke) [inline pandas/tests/frame/indexing/test_column_vs_index_tz.py:5]: Don't need an individual class for this. Writing the test as a function is sufficient.

Don't need a new file for just this test. This test can live in `pandas/tests/frame/test_query_eval.py`

File: pandas/tests/frame/indexing/test_column_vs_index_tz.py
Comment (mroeschke) [inline pandas/tests/frame/indexing/test_column_vs_index_tz.py:7]: As mentioned by @MarcoGorelli, you can just use the `tz_aware_fixture` here

File: pandas/tests/frame/indexing/test_column_vs_index_tz.py
Comment (mroeschke) [inline pandas/tests/frame/indexing/test_column_vs_index_tz.py:20]: Can you rename the variables so this reads `tm.assert_frame_equal(result, expected)`

File: pandas/tests/frame/indexing/test_column_vs_index_tz.py
Comment (mroeschke) [inline pandas/tests/frame/indexing/test_column_vs_index_tz.py:15]: Is this is the "expected" output, we'll want to construct this with a different method without query

```
expected = pd.DataFrame(...)
```

Comment (vipulrai91) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:15]: @mroeschke  ,    
I modified everything as per suggestion and the tests pass locally.
But here instead of constructing new "expected" DataFrame we would want to use the ones already generating by using query since "We are testing for result to be same at index and column level"
This is what was expected as per issue ? Right ?

Comment (mroeschke) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:15]: We want to make sure the "expected" DataFrame takes an independent path than query so if `query` breaks both "result" and "expected" don't change.

So want to make sure the result of:
`result = df.set_index('time').query('"2019-01-03 00:00:00+00" < time')`

gives a result of:
`expected = pd.DataFrame(...)`

Comment (vipulrai91) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:15]: When we create an expected dataframe , it passes for few tz and fails for others , bcoz the query string always expects the time zone to be "UTC" , query('"2019-01-03 00:00:**00+00" < time')**, and therefore there is shape mismatch  for example for "Asia/Kolkata" it is failing whereas it passes for "US/Eastern".

I assume we cannot change the query string because that is what we are testing for , in that case knowing the expected o/p when comparing to different tz will result in shape mismatch which is exactly happening now.
Any thoughts ?

```
def test_check_column_vs_index_tz_query(self, tz_aware_fixture):
    # https://github.com/pandas-dev/pandas/issues/29463

    tz = tz_aware_fixture
    expected = pd.DataFrame(
        {"time": pd.date_range(start="2019-01-04", freq="1d", periods=7, tz=tz)}
    ).set_index("time")

    df = pd.DataFrame(
        {"time": pd.date_range(start="2019-01-01", freq="1d", periods=10, tz=tz)}
    )
    result = df.set_index("time").query('"2019-01-03 00:00:00+00" < time')
    tm.assert_frame_equal(result, expected)
```

```
AssertionError: DataFrame are different

DataFrame shape mismatch
[left]:  (7, 0)
[right]: (8, 0)
```

Comment (mroeschke) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py:15]: Since we're just testing that we can query with a tz offset in the query string when the index is tz-aware, we can be with what the query string is.

You can just change the test to be then

```
    tz = tz_aware_fixture
    expected = pd.DataFrame(index=pd.date_range(start="2019-01-04", freq="1d", periods=10, tz=tz, name='time'))

    df = pd.DataFrame(index=pd.date_range(start="2019-01-01", freq="1d", periods=10, tz=tz, name=time))
    result = df.query('"2018-01-03 00:00:00+00" < time')
    tm.assert_frame_equal(result, expected)
```

Comment (vipulrai91) [reply pandas/tests/frame/indexing/test_column_vs_index_tz.py]: @mroeschke , Yes this works , I am now banging my head for not thinking this earlier.
Thank you very much , now all the tests pass.

As suggested , I have added the test in test_query_eval.py file.

File: pandas/tests/frame/test_query_eval.py
Comment (MarcoGorelli) [inline pandas/tests/frame/test_query_eval.py:712]: Like this, `time` is always greater than `2018-01-03 00:00:00+00` - I think @mroeschke was suggesting to use, say, `2019-01-04` in `expected`, `2019-01-01` in `df`, and then `2019-01-03` in `query`

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Nice one @vipulrai91 , looks better like this, just got a comment about the dates

File: pandas/tests/frame/test_query_eval.py
Comment (vipulrai91) [reply pandas/tests/frame/test_query_eval.py:712]: @MarcoGorelli  Thank you for the comment / appreciation.

If we use 2019-01-03 in query it results in earlier problem , because the shape of "expected" and "df" remains the same also their tz is changing , but in query the tz remains constant and when we query it to diff timezone their shape becomes inconsistent as the query tz  is always constant. (So it always result in +1 or -1 records and hence fails.)

Hence I think he suggested to use 2018-01-03 , as it will always result in consistent shape while checking tz-aware for all time zones.

Screenshot 2020-05-07 at 02 01 14

Comment (MarcoGorelli) [reply pandas/tests/frame/test_query_eval.py:712]: Ah sorry I'd misunderstood, thanks!

Comment (MarcoGorelli) [reply pandas/tests/frame/test_query_eval.py:712]: In that case, it seems that your index for both `df` and `expected` are identical - if so, can you make a variable to share them?

Comment (vipulrai91) [reply pandas/tests/frame/test_query_eval.py:707]: Done ! Thanks.

**Review by mroeschke — COMMENTED** [review]
LGTM. If you could merge master that will fix the failing CI

**Comment by vipulrai91** [comment]
> LGTM. If you could merge master that will fix the failing CI

Thanks. Merging now.

**Comment by vipulrai91** [comment]
> LGTM. If you could merge master that will fix the failing CI

@mroeschke 
I don't see merge option, also it says only those with write access can merge pull requests.

**Comment by vipulrai91** [comment]
> > LGTM. If you could merge master that will fix the failing CI
> 
> Thanks. Merging now.

 @mroeschke  All checks passes.

**Review by mroeschke — APPROVED** [review]
LGTM

**Review by MarcoGorelli — APPROVED** [review]

File: pandas/tests/frame/test_query_eval.py
Comment (jreback) [inline pandas/tests/frame/test_query_eval.py:709]: can you also test the working case here, e.g. ```result = df.reset_index().query(....)```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_query_eval.py
Comment (vipulrai91) [reply pandas/tests/frame/test_query_eval.py:713]: @jreback  Done. Please review.

Comment (MarcoGorelli) [reply pandas/tests/frame/test_query_eval.py:713]: @vipinkjonwal thanks for your updates

Like this, what's the difference between `df2.set_index("time")` and `df`?

I think Jeff meant
```python
result = df.reset_index().query('"2018-01-03 00:00:00+00" < time')
expected = pd.DataFrame(df_index)
tm.assert_frame_equal(result, expected)
```

Comment (vipulrai91) [reply pandas/tests/frame/test_query_eval.py:708]: @MarcoGorelli , I thought this was already working when the related issue was raised , I've updated the test with suggested changes.

**Review by mroeschke — COMMENTED** [review]
Sorry this PR hasn't been merged yet. Mind merging master into this branch?

After that, we can probably merge this in

**Comment by vipulrai91** [comment]
> Sorry this PR hasn't been merged yet. Mind merging master into this branch?
> 
> After that, we can probably merge this in

@mroeschke  done.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks for your patience @vipulrai91. Nice contribution!

**Comment by vipulrai91** [comment]
> Thanks for your patience @vipulrai91. Nice contribution!

Thank you @mroeschke @MarcoGorelli @jreback  for the help.
Looking forward to more contributions.

**Pull request merged**
